### PR TITLE
3.next - Integrate CookieCollection into Response

### PR DESCRIPTION
**Pull request by markstory**
Integrate the new Cookies and CookieCollection into the Response. All of the existing cookie related methods are backwards compatible, and `withCookie()` allows new style cookies as a parameter.

I've had to add more compatibility shims into Cookie and relax the expires parameter typing to allow integers. I found this preferable to coercing the existing cookie data into a DateTime just so it can be converted back into an integer.

Refs #10406

### CODE REVIEW

File: src/Http/Response.php [hunk 1]
```diff
@@ -462,6 +461,7 @@ public function __construct(array $options = [])
             $this->_contentType = $this->resolveType($options['type']);
         }
         $this->_setContentType();
+        $this->_cookies = new CookieCollection();
```
Comment (robertpustulka) [inline src/Http/Response.php]: Can we inject a collection from `$options`?

File: src/Http/Response.php [hunk 2]
```diff
@@ -1977,30 +1985,44 @@ public function cookie($options = null)
...
-            'domain' => '',
-            'secure' => false,
-            'httpOnly' => false
-        ];
-        $data += $defaults;
-        $data['name'] = $name;
 
         $new = clone $this;
-        $new->_cookies[$name] = $data;
+        $new->_cookies = $new->_cookies->add($cookie);
```
Comment (robertpustulka) [inline src/Http/Response.php]: I think a collection should be cloned as well. Adding a cookie to the new request's collection would also mutate current request's collection.

File: src/Http/Response.php (hunk 1 above)
Comment (markstory) [inline src/Http/Response.php]: Not right now. Responses don't take cookies in their constructor at this time.

File: src/Http/Cookie/Cookie.php [hunk 3]
```diff
@@ -251,7 +271,7 @@ public function getName()
      */
     protected function validateName($name)
     {
-        if (preg_match("/[=,; \t\r\n\013\014]/", $name)) {
+        if (preg_match("/[=,;\t\r\n\013\014]/", $name)) {
```
Comment (ravage84) [inline src/Http/Cookie/Cookie.php]: @markstory do you know of any documented list of invalid characters? If so, we could link to it with a ``@link``.

Comment (markstory) [inline src/Http/Cookie/Cookie.php]: The cookie RFCs have the accepted character lists.

Comment (ravage84) [inline src/Http/Cookie/Cookie.php]: If I interpreted that correctly:

> cookie-name       = token
> token             = <token, defined in [RFC2616], Section 2.2>

https://tools.ietf.org/html/rfc6265#section-4.1.1

So we could add:
``@link https://tools.ietf.org/html/rfc2616#section-2.2 Rules for naming cookies.``

Comment (markstory) [inline src/Http/Cookie/Cookie.php]: done.

**Pull request merged**